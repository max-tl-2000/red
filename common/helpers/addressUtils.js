/*
 * Copyright (c) 2022 Reva Technology Inc., all rights reserved.
 * Unauthorized copying of this file, via any medium is strictly prohibited
 * Licensed under the Elastic License 2.0; you may not use this file except
 * in compliance with the Elastic License 2.0.
 */

import { formatInventoryName } from '../inventory-helper';
import { isString } from './type-of';
const ZIP_CODE_SEPARATOR = ' ';

export const formatSimpleAddress = ({ addressLine1, addressLine2, city, state, zip, postalCode }) =>
  [addressLine1, addressLine2, city, state, zip || postalCode].filter(item => item).join(' ');

const toReadableString = (values, connector = ', ') => values.filter(item => item).join(connector);
export const formatShortAddress = ({ addressLine1, city }) => [addressLine1, city].filter(item => item).join(', ');

export const formatLongAddress = ({ addressLine1, city, state, zip, postalCode }) =>
  `${[addressLine1, city, state].filter(item => item).join(', ')}${ZIP_CODE_SEPARATOR}${zip || postalCode}`;

export const formatFullAddress = ({ addressLine1, addressLine2, city, state, zip, postalCode }) => ({
  fullAddressLine1: addressLine1,
  fullAddressLine2: addressLine2,
  fullAddressLine3: `${[city, state].filter(item => item).join(', ')}${ZIP_CODE_SEPARATOR}${zip || postalCode}`,
});
export const formatStateAndPostalCode = (state, postalCode) => toReadableString([state, postalCode], ' ');

const hasValidAddress = address => !!Object.keys(address || {}).length;

export const hasValidBuildingAddress = (building = {}) => hasValidAddress(building.address);

export const getUnitAddress = propertyName => (inventory = {}) => {
  if (!hasValidBuildingAddress(inventory.building)) return {};
  const COVE = 'cove';
  const SWPARKME = 'swparkme';
  const SHARON = 'sharon';
  const { city, state, postalCode } = inventory.building.address;
  let addressLine1 = inventory.building.address.addressLine1;
  let addressLine2 = inventory.building.address.addressLine2;

  if (propertyName === SHARON) {
    // sharon = <building.addressLine1>, <building.addressLine2>, <inventory.type> <building.name>-<inventory.name>, <building.city>, <building.state> <building.postalCode>""
    // For example - 350 Sharon Park Drive, Unit G-005, Menlo Park, CA 94025"
    addressLine2 = toReadableString([addressLine2, formatInventoryName(inventory, true)]);
  } else if (propertyName === COVE || propertyName === SWPARKME) {
    // <inventory.address>, <building.city>, <building.state> <building.postalCode>
    // For example -  111 Barbaree Way, Belvedere Tiburon, CA 94920
    addressLine1 = inventory.address;
    addressLine2 = '';
  } else {
    // lark, wood, shore = <building.addressLine1>, <building.addressLine2>, <inventory.type> <inventory.name>, <building.city>, <building.state> <building.postalCode>
    // For example - 2300 Lincoln village circle, Unit 164, Larkspur, CA 94536
    addressLine2 = toReadableString([addressLine2, formatInventoryName(inventory)]);
  }
  return {
    addressLine1,
    addressLine2,
    city,
    state,
    postalCode,
  };
};

export const formatUnitAddress = (inventory = {}) => {
  const { addressLine1: buildingAddressLine1, addressLine2: buildingAddressLine2, city: buildingCity, state: buildingState, postalCode: buildingPostalCode } =
    (inventory.building || {}).address || {};

  const { addressLine1: propertyAddressLine1, addressLine2: propertyAddressLine2, city: propertyCity, state: propertyState, postalCode: propertyPostalCode } =
    (inventory.property || {}).address || {};

  let address;
  if (inventory.inventoryAddress || (inventory.address && isString(inventory.address))) {
    // <inventory.address>, <building.city>, <building.state>, <building.postalCode>
    // eg. 111 Barbaree Way, Belvedere Tiburon, CA 94920
    const inventoryCity = buildingCity || propertyCity;
    const inventoryState = buildingState || propertyState;
    const inventoryPostalCode = buildingPostalCode || propertyPostalCode;

    address = [inventory.inventoryAddress || inventory.address, inventoryCity, formatStateAndPostalCode(inventoryState, inventoryPostalCode)];
  } else if (hasValidBuildingAddress(inventory.building)) {
    // <building.addressLine1>, <building.addressLine2>, <inventory.type> <inventory.name>, <building.city>, <building.state> <building.postalCode>
    // eg. 2300 Lincoln village circle, Unit 164, Larkspur, CA 94536
    address = [
      buildingAddressLine1,
      buildingAddressLine2,
      formatInventoryName(inventory),
      buildingCity,
      formatStateAndPostalCode(buildingState, buildingPostalCode),
    ];
  }

  // <property.addressLine1>, <property.addressLine2>, <inventory.type> <inventory.name>, <property.city>, <property.state> <property.postalCode>
  // eg. 2300 Lincoln village circle, Unit 164, Larkspur, CA 94536
  if (!address) {
    address = [
      propertyAddressLine1,
      propertyAddressLine2,
      formatInventoryName(inventory),
      propertyCity,
      formatStateAndPostalCode(propertyState, propertyPostalCode),
    ];
  }

  return toReadableString(address);
};

export const formatPropertyAddress = ({ addressLine1, addressLine2, city, state, postalCode }) =>
  toReadableString([addressLine1, addressLine2, city, formatStateAndPostalCode(state, postalCode)]);

export const formatBuildingAddress = inventory => {
  const buildingAddress = inventory.building.address;
  const propertyAddress = inventory.property.address;

  if (!hasValidBuildingAddress(inventory.building)) return formatPropertyAddress(propertyAddress);

  const { addressLine1, addressLine2 } = buildingAddress;
  const city = buildingAddress.city || propertyAddress.city;
  const state = buildingAddress.state || propertyAddress.state;
  const postalCode = buildingAddress.postalCode || propertyAddress.postalCode;

  return toReadableString([addressLine1, addressLine2, city, formatStateAndPostalCode(state, postalCode)]);
};
