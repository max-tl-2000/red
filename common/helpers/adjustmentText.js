/*
 * Copyright (c) 2022 Reva Technology Inc., all rights reserved.
 * Unauthorized copying of this file, via any medium is strictly prohibited
 * Licensed under the Elastic License 2.0; you may not use this file except
 * in compliance with the Elastic License 2.0.
 */

import { t } from 'i18next';
import { getFixedAmount } from './number';
import { formatMoney } from '../money-formatter';
import { USD } from '../currency';
import { DALTypes } from '../enums/DALTypes';

const shortPeriods = {
  year: 'y',
  week: 'w',
  month: 'm',
  day: 'd',
  hour: 'h',
};

const calculateAdjustmentValue = concession => {
  const { absoluteAdjustment, relativeAmount, amountVariableAdjustment, floorCeilingAmount } = concession;

  let adjustment = 0;

  if (amountVariableAdjustment > 0) {
    adjustment = amountVariableAdjustment;
  } else if (floorCeilingAmount > 0) {
    adjustment = floorCeilingAmount;
  } else if (relativeAmount) {
    adjustment = relativeAmount;
  } else if (parseFloat(absoluteAdjustment) !== 0) {
    adjustment = Math.abs(absoluteAdjustment);
  }

  return getFixedAmount(adjustment, 2);
};

const getShortPeriodFor = period => (period && shortPeriods[period]) || period;

export const adjustmentText = (concession, term) => {
  // *** This is where we need the override value
  const { recurringCount } = concession;

  const shortPeriod = getShortPeriodFor(term.period);
  const adjustmentValue = calculateAdjustmentValue(concession);
  const { result: formattedAdjustmentValue } = formatMoney({
    amount: adjustmentValue,
    currency: USD.code,
  });
  const adjustment = `(${formattedAdjustmentValue}${t('PER_PERIOD', {
    period: shortPeriod,
  })}`;

  const count = recurringCount > 0 ? recurringCount : term.termLength;

  return `${adjustment} ${t('FOR_PERIOD', {
    length: count,
    period: shortPeriod,
  })})`;
};

export const getStartingAtPriceText = (marketRent, term = { termLength: 12, period: DALTypes.LeasePeriod.MONTH }, appendPeriodInfo = true) => {
  const { integerPart, decimalPart } = formatMoney({
    amount: marketRent,
    currency: USD.code,
  });
  const startingAtPriceText = t('STARTING_AT', {
    integerPart,
    decimalPart,
  });
  if (!appendPeriodInfo) return startingAtPriceText;

  const shortPeriod = getShortPeriodFor(term.period);
  const shortPreiordText = t('PER_PERIOD', { period: shortPeriod });
  const forPeriodText = `${shortPreiordText} ${t('FOR_PERIOD', {
    length: term.termLength,
    period: shortPeriod,
  })}`;

  return `${startingAtPriceText}${forPeriodText}`;
};
