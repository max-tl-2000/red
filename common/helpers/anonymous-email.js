/*
 * Copyright (c) 2022 Reva Technology Inc., all rights reserved.
 * Unauthorized copying of this file, via any medium is strictly prohibited
 * Licensed under the Elastic License 2.0; you may not use this file except
 * in compliance with the Elastic License 2.0.
 */

const anonymizedEmails = {
  'zlead.co': [/^.{26}@zlead.co$/],
  'reply.craigslist.org': [/^.{32}@reply.craigslist.org$/],
  'renter.apartmentlist.com': [/^.+\..+\..{5}@renter.apartmentlist.com$/],
  'messaging.yelp.com': [/^reply\+([^@]*)@messaging.yelp.com$/],
  'message.my.apartmentguide.com': [/^reply-([^@]*)@message.my.apartmentguide.com$/],
  'email.rent.com': [/^reply-([^@]*)@email.rent.com$/],
  'abodoapts.com': [/^abodoleads\+([^@]*)@abodoapts.com$/],
};

const isAnonymizedEmail = (email, searchExpressions = []) => searchExpressions.some(expression => new RegExp(expression, 'i').test(email));

export const isAnonymousEmail = email => email && Object.keys(anonymizedEmails).some(provider => isAnonymizedEmail(email, anonymizedEmails[provider]));
