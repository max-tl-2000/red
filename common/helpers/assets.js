/*
 * Copyright (c) 2022 Reva Technology Inc., all rights reserved.
 * Unauthorized copying of this file, via any medium is strictly prohibited
 * Licensed under the Elastic License 2.0; you may not use this file except
 * in compliance with the Elastic License 2.0.
 */

const getUrlParams = ({ tenantName, cloudEnv }, apiToken) => `env=${cloudEnv}&tenant=${tenantName}&api-token=${apiToken}`;

export const formatStaticAssetUrl = (ctx, assetName, assetType) => {
  const { isPublicEnv, domainSuffix, reverseProxyUrl, rpImageToken, cloudEnv } = ctx;

  if (!domainSuffix || !reverseProxyUrl || !rpImageToken || !cloudEnv || !assetName || !assetType) return '';

  const staticAssetsBaseUrl = isPublicEnv ? `https://static.${domainSuffix}/images` : `${reverseProxyUrl}/images`;
  const assetUrl = `${staticAssetsBaseUrl}/${assetType.toLowerCase()}/${assetName}`;

  return isPublicEnv ? assetUrl : `${assetUrl}?${getUrlParams({ cloudEnv, tenantName: 'static' }, rpImageToken)}`;
};
