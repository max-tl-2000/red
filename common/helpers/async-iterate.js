/*
 * Copyright (c) 2022 Reva Technology Inc., all rights reserved.
 * Unauthorized copying of this file, via any medium is strictly prohibited
 * Licensed under the Elastic License 2.0; you may not use this file except
 * in compliance with the Elastic License 2.0.
 */

export const asyncIterate = (items, { itemCb, done, delay = 1000 } = {}) => {
  if (!Array.isArray(items)) {
    throw new Error('first argument must be an array');
  }
  if (typeof itemCb !== 'function') {
    throw new Error('itemCb callback is required');
  }

  const len = items.length;
  let index = 0;

  if (len === 0) {
    done && done();
    return;
  }

  const process = () => {
    const item = items[index];
    itemCb({ item, index }, () => {
      index++;
      if (index === len) {
        done && done();
        return;
      }
      setTimeout(process, delay);
    });
  };
  process();
};

export const iterateOverArray = (items = [], args) => {
  const { chunkSize = 100, onChunk, done } = args;
  const { length } = items;

  const totalGroups = Math.ceil(length / chunkSize);
  let processed = 0;
  let cancelled = false;
  let iterationDone = false;

  const process = () => {
    if (cancelled) {
      return;
    }

    if (processed === totalGroups) {
      iterationDone = true;
      done && done();
      return;
    }

    const from = processed * chunkSize;
    const to = from + chunkSize;
    const arr = items.slice(from, to);

    setTimeout(() => {
      if (cancelled) return;
      onChunk &&
        onChunk({ arr, from, to: to - 1 }, () => {
          processed++;
          setTimeout(process);
        });
    });
  };

  process();

  return () => {
    if (iterationDone || cancelled) return;
    cancelled = true;
  };
};
