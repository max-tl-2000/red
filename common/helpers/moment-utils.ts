/*
 * Copyright (c) 2022 Reva Technology Inc., all rights reserved.
 * Unauthorized copying of this file, via any medium is strictly prohibited
 * Licensed under the Elastic License 2.0; you may not use this file except
 * in compliance with the Elastic License 2.0.
 */

import moment, { Moment } from 'moment-timezone'; // eslint-disable-line red/no-moment
import * as baseMoment from 'moment'; // eslint-disable-line red/no-moment
import { extendMoment } from 'moment-range';
import nullish from './nullish';
import { isString } from './type-of';
import { isNode } from './globals';

export type MomentType = Moment | string | Date;

interface IIsSameDayOptions {
  timezone?: string;
}

interface ILogWarningsArgs {
  msg?: any[];
}

interface IToMomentArgs {
  parseFormat?: string;
  timezone?: string;
  strict?: boolean;
}

interface IFormatMomentArgs {
  format?: string;
  timezone?: string;
  includeZone?: boolean;
  includeZoneStrict?: boolean;
}

interface INowArgs {
  timezone?: string;
}

interface IParseASInTimezoneArgs {
  format?: string;
  timezone?: string;
}

const runningInNode = isNode();

const momentFn = moment;
const { duration, max: momentMax, now: momentNow } = momentFn;

export const DATE_ISO_FORMAT = 'YYYY-MM-DD';

// this format is not correct the UTC formats with `Z` don't include +z
export const DATE_TIME_ISO_FORMAT = 'YYYY-MM-DDTHH:mm:ss+z';

export const findLocalTimezone = (): string => momentFn.tz.guess();
export const isValidTimezone = (tz: string): boolean => !!momentFn.tz.zone(tz);

let momentUtilsFixedNow;

const logWarnings = ({ msg = [] }: ILogWarningsArgs = {}) => {
  const isProduction = process.env.NODE_ENV === 'production';

  if (isProduction || !msg || msg.length === 0) {
    return;
  }

  const isDevelopment = process.env.NODE_ENV === 'development';

  console.warn(...msg);

  if (isDevelopment && runningInNode) {
    const err = new Error();
    const LINES_TO_SKIP = 2;
    // we remove the 2 first lines which are meaningless for us
    // and also the lines that contains stack info from teh asyncGenerator babel transform
    const stackWithMeaningFulInfo = (err.stack || '')
      .split(/\n/)
      .slice(LINES_TO_SKIP)
      .filter(line => !line.match(/asyncToGenerator\.js/))
      .join('\n');

    console.warn(stackWithMeaningFulInfo);
  }
};

const checkIfDateIsISO = (date: MomentType): void => {
  const isProduction = process.env.NODE_ENV === 'production';

  if (isProduction || !isString(date)) {
    return;
  }

  date = date as string; // cast needed to make typescript happy

  const ISO_8601_REGEX_FULL = /^\d{4}-\d\d-\d\dT\d\d:\d\d:\d\d(\.\d+)?(([+-]\d\d:\d\d)|Z)?$/i;
  const ISO_8601_REGEX_PERMISSIVE = /^\d{4}(-\d\d(-\d\d(T\d\d:\d\d(:\d\d)?(\.\d+)?(([+-]\d\d:\d\d)|Z)?)?)?)?$/i;

  if (date.match(ISO_8601_REGEX_FULL)) {
    return;
  }

  if (date.match(ISO_8601_REGEX_PERMISSIVE)) {
    // if it matches the permissive then we just inform it failed the full ISO validation
    logWarnings({ msg: ['>> `date` parameter is not in full ISO format', date] });
    return;
  }

  logWarnings({ msg: ['>> date parameter is not in any ISO format', date] });
};

export const toMoment = (date: MomentType, { parseFormat, timezone, strict = true }: IToMomentArgs = { strict: true }): Moment => {
  if (nullish(date) || date === '') {
    logWarnings({
      msg: ['>> date parameter is nullish or empty', date, '\n', '>> this will produce an invalid moment that will be formatted as empty string'],
    });

    const m = momentFn(null!);
    m.format = () => '';
    return m; // just return an invalid moment object
  }

  if (parseFormat !== undefined && typeof parseFormat !== 'string') {
    throw new TypeError('`parseFormat` must be a string');
  }

  checkIfDateIsISO(date);

  if (timezone !== undefined && !isString(timezone)) {
    throw new TypeError('timezone must be a string');
  }

  let dateAsMoment = parseFormat ? momentFn(date, parseFormat, strict) : momentFn(date);

  if (timezone) {
    dateAsMoment = dateAsMoment.tz(timezone);
  }

  return dateAsMoment;
};

export const formatMoment = (
  momentObj: Moment,
  { format, timezone, includeZone = true, includeZoneStrict = false }: IFormatMomentArgs = { includeZone: true, includeZoneStrict: false },
): string => {
  if (!timezone) {
    logWarnings({ msg: ['[formatMoment] called without timezone', momentObj] });
  }

  const m = momentObj ? toMoment(momentObj, { timezone }) : undefined;

  if (!m || !m.isValid()) {
    logWarnings({ msg: ['[formatMoment] attempt to format an invalid moment', momentObj] });
    return '';
  }

  const useAbbr = timezone && includeZone && timezone !== findLocalTimezone();

  const zone = useAbbr || includeZoneStrict ? ` ${m.zoneAbbr()}` : '';
  return `${m.format(format)}${zone}`;
};

export const now = ({ timezone = findLocalTimezone() }: INowArgs = { timezone: findLocalTimezone() }) => {
  if (momentUtilsFixedNow) {
    return momentUtilsFixedNow;
  }

  if (timezone !== undefined && typeof timezone !== 'string') {
    throw new TypeError('timezone must be a string');
  }

  let m = momentFn();
  if (timezone) {
    m = m.tz(timezone);
  }
  return m;
};

export const useFixedDateTimeForMomentUtilsNow = dateTime => {
  momentUtilsFixedNow = dateTime;

  return () => {
    momentUtilsFixedNow = undefined;
  };
};

export const parseAsInTimezone = (str: string, { format, timezone }: IParseASInTimezoneArgs = {}): Moment => {
  const args = [str];

  if (format) {
    args.push(format);
  }

  if (timezone) {
    args.push(timezone);
  } else {
    logWarnings({ msg: ['[parseAsInTimezone] error: missing `timezone`. In future versions this will throw an error'] });
  }

  // @ts-ignore: spread will work just fine in this case
  return momentFn.tz(...args);
};

export const isMoment = (m: any): boolean => momentFn.isMoment(m);

export const isSameDay = (date1, date2, { timezone }: IIsSameDayOptions = {}): boolean => {
  if (!timezone) {
    logWarnings({ msg: ['[isSameDay] called without timezone'] });
  }

  date1 = toMoment(date1, { timezone });
  date2 = toMoment(date2, { timezone });

  return !date1.isValid() || !date2.isValid() ? false : date1.isSame(date2, 'day');
};

export const isValidMoment = (m: any): boolean => m && isMoment(m) && m.isValid();

export const setDefaultTZ = (tz: string) => momentFn.tz.setDefault(tz);

export const getDurationBetweenMoments = (startDate: MomentType, endDate: MomentType): string => {
  const millisecondDifference = toMoment(endDate).diff(startDate);
  return moment.utc(millisecondDifference).format('mm:ss');
};

export const rangeMoment = extendMoment(baseMoment);

export const getTimezoneNames = (): Array<string> => momentFn.tz.names();

export const stringRepresentsADateWithoutTime = date => {
  if (typeof date !== 'string') {
    return false;
  }
  const POSIBLE_DATE = /^\d{4}-\d{1,2}-\d{1,2}$/g;
  return !!date.match(POSIBLE_DATE);
};

export const isWithinLastNDays = ({ date, days, timezone }) => toMoment(date, { timezone }).isSameOrAfter(now().add(days, 'days'), 'day');

export { duration, momentMax, momentNow };

export const today = ({ timezone = findLocalTimezone() }: INowArgs = { timezone: findLocalTimezone() }) =>
  now({ timezone }).set({ hour: 0, minute: 0, second: 0, millisecond: 0 });
