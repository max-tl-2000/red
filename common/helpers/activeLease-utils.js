/*
 * Copyright (c) 2022 Reva Technology Inc., all rights reserved.
 * Unauthorized copying of this file, via any medium is strictly prohibited
 * Licensed under the Elastic License 2.0; you may not use this file except
 * in compliance with the Elastic License 2.0.
 */

import { t } from 'i18next';
import { toMoment } from './moment-utils';

export const getActiveLeaseEndDate = activeLeaseWorkflowData => {
  const leaseData = activeLeaseWorkflowData?.leaseData;

  return leaseData?.computedExtensionEndDate || leaseData?.leaseEndDate;
};

export const enhanceChargesWithAdjustedDates = (activeLeaseWorkflowData, timezone) => {
  const { recurringCharges, concessions } = activeLeaseWorkflowData;
  const formattedRecurringCharges = (recurringCharges || []).map(r => {
    const startDate = toMoment(r.startDate, { timezone }).add(2, 'hours').toISOString();
    const endDate = r.endDate ? toMoment(r.endDate, { timezone }).add(2, 'hours').toISOString() : null;
    return {
      ...r,
      startDate,
      endDate,
    };
  });

  const formattedConcessions = (concessions || []).map(c => {
    const startDate = toMoment(c?.startDate, { timezone }).add(2, 'hours').toISOString();
    const endDate = c.endDate ? toMoment(c.endDate, { timezone }).add(2, 'hours').toISOString() : null;
    return {
      ...c,
      startDate,
      endDate,
    };
  });

  return {
    ...activeLeaseWorkflowData,
    recurringCharges: formattedRecurringCharges,
    concessions: formattedConcessions,
  };
};

export const getFormattedLeaseEndDate = (formattedEndDate, daysBeforeLeaseEnds) => {
  const endDate = t('DATE', { date: formattedEndDate });

  if (daysBeforeLeaseEnds < -1) return `${endDate} ${t('DAYS_AGO', { noOfDays: -daysBeforeLeaseEnds })}`;
  if (daysBeforeLeaseEnds === -1) return `${endDate} ${t('REMAINING_YESTERDAY')}`;
  if (daysBeforeLeaseEnds === 0) return `${endDate} ${t('REMAINING_TODAY')}`;
  if (daysBeforeLeaseEnds === 1) return `${endDate} ${t('REMAINING_TOMORROW')}`;
  if (daysBeforeLeaseEnds > 1) return `${endDate} ${t('REMAINING_DAYS', { noOfDays: daysBeforeLeaseEnds })}`;
  return `${endDate}`;
};
