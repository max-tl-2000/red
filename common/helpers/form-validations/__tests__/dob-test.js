/*
 * Copyright (c) 2022 Reva Technology Inc., all rights reserved.
 * Unauthorized copying of this file, via any medium is strictly prohibited
 * Licensed under the Elastic License 2.0; you may not use this file except
 * in compliance with the Elastic License 2.0.
 */

import * as momentUtils from 'helpers/moment-utils';
import { toMoment } from '../../moment-utils';

const { mockModules } = require('test-helpers/mocker').default(jest);

describe('isValidDOB', () => {
  let isValidDOB;
  beforeEach(() => {
    mockModules({
      i18next: {
        t: (...args) => args, // we just want to log what we receive to compare the values
      },
      '../../../../common/helpers/moment-utils': {
        ...momentUtils,
        now: () => toMoment('07/01/2017', { parseFormat: 'MM/DD/YYYY' }), // mock a current now moment set to a known date
      },
    });
    isValidDOB = require('../dob').isValidDOB; // eslint-disable-line
  });

  describe('when only a partial value of the date is entered', () => {
    it('should show the error message accordingly', () => {
      const res = isValidDOB({ value: '01' });
      expect(res).toMatchSnapshot();
    });
  });

  describe('when a valid value is entered', () => {
    it('should return true', () => {
      const res = isValidDOB({ value: '01/01/1979' });
      expect(res).toEqual(true);
    });
  });

  describe('when a date is before the min valid year', () => {
    it('should produce an error message and show the valid range', () => {
      const res = isValidDOB({ value: '01/01/1111' });
      expect(res).toMatchSnapshot();
    });
  });

  describe('when the applicant is not 18 years old', () => {
    it('should produce an error telling the user he needs to be at least 18', () => {
      const res = isValidDOB({ value: '01/01/2002' });
      expect(res).toMatchSnapshot();
    });
  });
});
