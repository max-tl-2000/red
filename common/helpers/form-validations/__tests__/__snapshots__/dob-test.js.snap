// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`isValidDOB when a date is before the min valid year should produce an error message and show the valid range 1`] = `
Object {
  "error": Array [
    "INVALID_DATE_RANGE",
    Object {
      "max": "07/01/1999",
      "min": 1900,
    },
  ],
}
`;

exports[`isValidDOB when only a partial value of the date is entered should show the error message accordingly 1`] = `
Object {
  "error": Array [
    "INVALID_DATE_FORMAT",
    Object {
      "format": "MM/DD/YYYY",
    },
  ],
}
`;

exports[`isValidDOB when the applicant is not 18 years old should produce an error telling the user he needs to be at least 18 1`] = `
Object {
  "error": Array [
    "YOU_MIGHT_BE_MIN_AGE_OR_OLDER",
    Object {
      "minAge": 18,
    },
  ],
}
`;
